### PR TITLE
[TGUI] Photocopier Refresh

### PR DESCRIPTION
**Pull request by AyIong**
## About The Pull Request
- New photocopier UI, more informative, user-friendly, with search and using TS
- You can now print several copies of blanks at once. I personally don't understand why limit was set to 1 form to print. It is easier to print the required number of copies at once than print 1 copy and copy them immediately
- Fixed broken Button ellipsis, by additing `display: block` to ellipsis span class

| Before | After - Default | After - All buttons (With photo inserted) |
| - | - | - | 
| image | image | image |

## Why It's Good For The Game
Better UX
Full blank name instead of code, who tf remember what blank behind code?
Search

## Changelog

:cl:
qol: Photocopier updated their UI
/:cl:

### CODE REVIEW

**Comment by MrMelbert** [comment]
I like the changes on the left but, well, I don't think you need a massive list of every blank on the right side of the screen

**Comment by AyIong** [comment]
> I like the changes on the left but, well, I don't think you need a massive list of every blank on the right side of the screen shown at all times

Maybe, but I think it's better this way, as you can quickly select the form you want without having to make unnecessary clicks
But I have to admit it takes up too much space, do you think it's better to hide it if no category is selected?

**Comment by MrMelbert** [comment]
You think you can make it so it starts condensed and opening the blanks expands the window? 

I forget if it's possible in tgui (I think it is?)

**Comment by AyIong** [comment]
> You think you can make it so it starts condensed and opening the blanks expands the window?
> 
> I forget if it's possible in tgui (I think it is?)

odUHQZuYEh

**Comment by MrMelbert** [comment]
(Also maybe rename the section from `Categories` to `Templates`?)

**Comment by AyIong** [comment]
> (Also maybe rename the section from `Categories` to `Templates`?)

But that's literally categories

**Comment by MrMelbert** [comment]
it's not immediately obvious what the categories "means"

**Comment by AyIong** [comment]
> it's not immediately obvious what the categories "means" 
> 
> if I'm just trying to photocopy a paper how is "categories" relevant to me

Makes sense, i'll change it tomorrow back to "Blanks"
Thanks for feedback <3

File: tgui/packages/tgui/interfaces/Photocopier.tsx [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+              </Stack>
+            </Stack.Item>
+          )}
+        </Stack>
+      </Window.Content>
+    </Window>
+  );
+};
+
+const Status = ({ selectedBlank }) => {
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Photocopier.tsx:88]: prefer we don't get into this pattern bc not only is it `any` type but it's also assigned as mutable `let selectedBlank` when you destructure inline like this. `props: PropTypeName` pls

File: tgui/packages/tgui/interfaces/Photocopier.tsx [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+          icon="times"
+          tooltip="Close category"
+          disabled={!selectedCategory}
+          onClick={() => {
+            setSelectedCategory('');
+          }}
+        />
+      }
+    >
+      <Stack fill vertical>
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Photocopier.tsx:310]: pls use candystripe :)

File: tgui/packages/tgui/interfaces/Photocopier.tsx [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+const Blanks = ({ selectedCategory, selectedBlank, setSelectedBlank }) => {
+  const { act, data } = useBackend<Data>();
+  const { blanks } = data;
+
+  const [searchText, setSearchText] = useState('');
+  const sortedBlanks = sortBy(blanks || [], (blank) => blank.name);
+  const visibleBlanks = searchText
+    ? sortedBlanks.filter((blank) =>
+        blank.name.toLowerCase().includes(searchText.toLowerCase()),
+      )
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Photocopier.tsx:350]: we have a `createSearch` for this

File: tgui/packages/tgui/interfaces/Photocopier.tsx [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+    </Section>
+  );
+};
+
+const Blanks = ({ selectedCategory, selectedBlank, setSelectedBlank }) => {
+  const { act, data } = useBackend<Data>();
+  const { blanks } = data;
+
+  const [searchText, setSearchText] = useState('');
+  const sortedBlanks = sortBy(blanks || [], (blank) => blank.name);
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Photocopier.tsx:346]: please just use the std lib `.sort`, see #83161

File: tgui/packages/tgui/interfaces/Photocopier.tsx [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        <Input
+          width={8.75}
+          value={searchText}
+          placeholder="Search blank..."
+          onInput={(e, value) => setSearchText(value)}
+        />
+      }
+    >
+      {visibleBlanks.map((blank) => (
+        <Stack.Item key={blank.code} width="100%">
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Photocopier.tsx:370]: candystripe here as well pls

File: tgui/packages/tgui/interfaces/Photocopier.tsx [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+                icon="copy"
+                disabled={!has_item}
+                onClick={() => act('make_copy')}
+              >
+                Copy
+              </Button>
+            </Stack.Item>
+          </Stack>
+        </Stack.Item>
+        {!!has_item && is_photo && (
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Photocopier.tsx:241]:
```suggestion
        {!!has_item && !!is_photo && (
```

**Comment by AyIong** [comment]
With candystripe
image

File: tgui/packages/tgui/interfaces/Photocopier.tsx (hunk 1 above)
Comment (jlsnow301) [reply tgui/packages/tgui/interfaces/Photocopier.tsx]: sorry, I wasn't specific enough here. I mean inlining the destructure statement itself. this assigns it as `let` 
```tsx
const Categories = ({
  selectedCategory,
  setSelectedCategory,
}: {
  selectedCategory: string | null;
  setSelectedCategory: (category: string) => void;
}
```
would become
```tsx
type CategoryProps = {
  selectedCategory: string | null;
  setSelectedCategory: (category: string) => void;
}

const Categories = (props: CategoryProps) => {
  const { selectedCategory, setSelectedCategory } = props;
```

Comment (AyIong) [reply tgui/packages/tgui/interfaces/Photocopier.tsx]: Oh.. well, thanks, i'll do it tomorrow <3
Just looked at other UI's, and it was like i did

File: tgui/packages/tgui/interfaces/Photocopier.tsx [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+
+const Blanks = (props: BlanksProps) => {
+  const { act, data } = useBackend<Data>();
+  const { selectedCategory, selectedBlank, setSelectedBlank } = props;
+  const { blanks } = data;
+
+  const [searchText, setSearchText] = useState('');
+  const search = createSearch<Blank>(searchText, (blank: Blank) => blank.name);
+  const sortedBlanks = (blanks || []).sort((a, b) =>
+    a.name.localeCompare(b.name),
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Photocopier.tsx:373]: please don't use this, it's much slower. use just `a - b`

File: tgui/packages/tgui/interfaces/Photocopier.tsx [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+  setSelectedBlank: (blank: string) => void;
+};
+
+const Blanks = (props: BlanksProps) => {
+  const { act, data } = useBackend<Data>();
+  const { selectedCategory, selectedBlank, setSelectedBlank } = props;
+  const { blanks } = data;
+
+  const [searchText, setSearchText] = useState('');
+  const search = createSearch<Blank>(searchText, (blank: Blank) => blank.name);
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/Photocopier.tsx:371]:
```suggestion
  const search = createSearch(searchText, (blank: Blank) => blank.name);
```

**Pull request merged**